File: app/src/main/java/com/etesync/syncadapter/ui/JournalItemActivity.java
package com.etesync.syncadapter.ui;

import android.content.Context;
import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;
import android.support.v4.view.ViewPager;
import android.text.format.DateFormat;
import android.text.format.DateUtils;
import android.text.format.Time;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.etesync.syncadapter.App;
import com.etesync.syncadapter.Constants;
import com.etesync.syncadapter.R;
import com.etesync.syncadapter.model.CollectionInfo;
import com.etesync.syncadapter.model.JournalEntity;
import com.etesync.syncadapter.model.SyncEntry;

import net.fortuna.ical4j.model.component.VAlarm;
import net.fortuna.ical4j.model.property.Attendee;

import org.apache.commons.codec.Charsets;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.Formatter;
import java.util.Locale;

import at.bitfire.ical4android.Event;
import at.bitfire.ical4android.InvalidCalendarException;
import io.requery.Persistable;
import io.requery.sql.EntityDataStore;

import static com.etesync.syncadapter.ui.journalviewer.ListEntriesFragment.setJournalEntryView;

public class JournalItemActivity extends BaseActivity implements Refreshable {
    private static final String KEY_SYNC_ENTRY = "syncEntry";
    private JournalEntity journalEntity;
    protected CollectionInfo info;
    private SyncEntry syncEntry;

    public static Intent newIntent(Context context, CollectionInfo info, SyncEntry syncEntry) {
        Intent intent = new Intent(context, JournalItemActivity.class);
        intent.putExtra(Constants.KEY_COLLECTION_INFO, info);
        intent.putExtra(KEY_SYNC_ENTRY, syncEntry);
        return intent;
    }

    @Override
    public void refresh() {
        EntityDataStore<Persistable> data = ((App) getApplicationContext()).getData();

        journalEntity = JournalEntity.fetch(data, info.getServiceEntity(data), info.uid);
        if ((journalEntity == null) || journalEntity.isDeleted()) {
            finish();
            return;
        }

        info = journalEntity.getInfo();

        setTitle(info.displayName);

        setJournalEntryView(findViewById(R.id.journal_list_item), info, syncEntry);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.journal_item_activity);

        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        info = (CollectionInfo) getIntent().getExtras().getSerializable(Constants.KEY_COLLECTION_INFO);
        syncEntry = (SyncEntry) getIntent().getExtras().getSerializable(KEY_SYNC_ENTRY);

        refresh();

        ViewPager viewPager = (ViewPager) findViewById(R.id.viewpager);
        viewPager.setAdapter(new TabsAdapter(getSupportFragmentManager(), info, syncEntry));

        TabLayout tabLayout = (TabLayout) findViewById(R.id.tabs);
        tabLayout.setupWithViewPager(viewPager);
    }

    private static class TabsAdapter extends FragmentPagerAdapter {
        private CollectionInfo info;
        private SyncEntry syncEntry;
        public TabsAdapter(FragmentManager fm, CollectionInfo info, SyncEntry syncEntry) {
            super(fm);
            this.info = info;
            this.syncEntry = syncEntry;
        }

        @Override
        public int getCount() {
            // FIXME: Make it depend on info type (only have non-raw for known types)
            return 2;
        }

        @Override
        public CharSequence getPageTitle(int position) {
            // FIXME: use string resources
            if (position == 0) {
                return "Main";
            } else {
                return "Raw";
            }
        }

        @Override
        public Fragment getItem(int position) {
            if (position == 0) {
                return EventFragment.newInstance(info, syncEntry);
            } else {
                return TextFragment.newInstance(syncEntry);
            }
        }
    }

    public static class TextFragment extends Fragment {
        public static TextFragment newInstance(SyncEntry syncEntry) {
            TextFragment frag = new TextFragment();
            Bundle args = new Bundle(1);
            args.putSerializable(KEY_SYNC_ENTRY, syncEntry);
            frag.setArguments(args);
            return frag;
        }

        @Override
        public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
            View v = inflater.inflate(R.layout.text_fragment, container, false);

            TextView tv = (TextView) v.findViewById(R.id.content);

            SyncEntry syncEntry = (SyncEntry) getArguments().getSerializable(KEY_SYNC_ENTRY);
            tv.setText(syncEntry.getContent());

            return v;
        }
    }

    public static class EventFragment extends Fragment {
        CollectionInfo info;
        SyncEntry syncEntry;

        public static EventFragment newInstance(CollectionInfo info, SyncEntry syncEntry) {
            EventFragment frag = new EventFragment();
            Bundle args = new Bundle(1);
            args.putSerializable(Constants.KEY_COLLECTION_INFO, info);
            args.putSerializable(KEY_SYNC_ENTRY, syncEntry);
            frag.setArguments(args);
            return frag;
        }

        @Override
        public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
            View v = inflater.inflate(R.layout.event_info, container, false);

            info = (CollectionInfo) getArguments().getSerializable(Constants.KEY_COLLECTION_INFO);
            syncEntry = (SyncEntry) getArguments().getSerializable(KEY_SYNC_ENTRY);

            new LoadEventTask(v).execute();

            return v;
        }

        private class LoadEventTask extends AsyncTask<Void, Void, Event> {
            View view;
            LoadEventTask(View v) {
                super();
                view = v;
            }
            @Override
            protected Event doInBackground(Void... aVoids) {
                InputStream is = new ByteArrayInputStream(syncEntry.getContent().getBytes(Charsets.UTF_8));

                try {
                    Event event = Event.fromStream(is, Charsets.UTF_8, null)[0];
                    return event;
                } catch (InvalidCalendarException e) {
                    e.printStackTrace();
                } catch (IOException e) {
                    e.printStackTrace();
                }

                return null;
            }

            @Override
            protected void onPostExecute(Event event) {
                final View loader = view.findViewById(R.id.event_info_loading_msg);
                loader.setVisibility(View.GONE);
                final View contentContainer = view.findViewById(R.id.event_info_scroll_view);
                contentContainer.setVisibility(View.VISIBLE);

                setTextViewText(view, R.id.title, event.summary);

                setTextViewText(view, R.id.when_datetime, getDisplayedDatetime(event.dtStart.getDate().getTime(), event.dtEnd.getDate().getTime(), event.isAllDay(), getContext()));

                setTextViewText(view, R.id.where, event.location);

                if (event.organizer != null) {
                    TextView tv = (TextView) view.findViewById(R.id.organizer);
                    tv.setText(event.organizer.getCalAddress().toString().replaceFirst("mailto:", ""));
                } else {
                    View organizer = view.findViewById(R.id.organizer_container);
                    organizer.setVisibility(View.GONE);
                }

                setTextViewText(view, R.id.description, event.description);

                boolean first = true;
                StringBuilder sb = new StringBuilder();
                for (Attendee attendee : event.attendees) {
                    if (first) {
                        first = false;
                        sb.append("Attendees: ");
                    } else {
                        sb.append(", ");
                    }
                    sb.append(attendee.getCalAddress().toString().replaceFirst("mailto:", ""));
                }
                setTextViewText(view, R.id.attendees, sb.toString());

                first = true;
                sb = new StringBuilder();
                for (VAlarm alarm : event.alarms) {
                    if (first) {
                        first = false;
                        sb.append("Reminders: ");
                    } else {
                        sb.append(", ");
                    }
                    sb.append(alarm.getTrigger().getValue());
                }
                setTextViewText(view, R.id.reminders, sb.toString());
            }
        }

        private static void setTextViewText(View parent, int id, String text) {
            TextView tv = (TextView) parent.findViewById(id);
            if (text == null) {
                tv.setVisibility(View.GONE);
            } else {
                tv.setText(text);
            }
        }

        public static String getDisplayedDatetime(long startMillis, long endMillis, boolean allDay, Context context) {
            // Configure date/time formatting.
            int flagsDate = DateUtils.FORMAT_SHOW_DATE | DateUtils.FORMAT_SHOW_WEEKDAY;
            int flagsTime = DateUtils.FORMAT_SHOW_TIME;
            if (DateFormat.is24HourFormat(context)) {
                flagsTime |= DateUtils.FORMAT_24HOUR;
            }

            String datetimeString = null;
            if (allDay) {
                // For multi-day allday events or single-day all-day events that are not
                // today or tomorrow, use framework formatter.
                Formatter f = new Formatter(new StringBuilder(50), Locale.getDefault());
                datetimeString = DateUtils.formatDateRange(context, f, startMillis,
                        endMillis, flagsDate, Time.TIMEZONE_UTC).toString();
            } else {
                // For multiday events, shorten day/month names.
                // Example format: "Fri Apr 6, 5:00pm - Sun, Apr 8, 6:00pm"
                int flagsDatetime = flagsDate | flagsTime | DateUtils.FORMAT_ABBREV_MONTH |
                        DateUtils.FORMAT_ABBREV_WEEKDAY;
                datetimeString = DateUtils.formatDateRange(context, startMillis, endMillis,
                        flagsDatetime);
            }
            return datetimeString;
        }
    }

    @Override
    protected void onResume() {
        super.onResume();
        refresh();
    }
}

File: app/src/main/java/com/etesync/syncadapter/ui/JournalItemActivity.kt
package com.etesync.syncadapter.ui

import android.accounts.Account
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.text.format.DateFormat
import android.text.format.DateUtils
import android.view.*
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter
import androidx.viewpager.widget.ViewPager
import at.bitfire.ical4android.Event
import at.bitfire.ical4android.InvalidCalendarException
import at.bitfire.ical4android.Task
import at.bitfire.vcard4android.Contact
import com.etesync.syncadapter.App
import com.etesync.syncadapter.Constants
import com.etesync.syncadapter.R
import com.etesync.syncadapter.model.CollectionInfo
import com.etesync.syncadapter.model.JournalEntity
import com.etesync.syncadapter.model.SyncEntry
import com.etesync.syncadapter.ui.journalviewer.ListEntriesFragment.Companion.setJournalEntryView
import com.etesync.syncadapter.utils.EventEmailInvitation
import com.google.android.material.tabs.TabLayout
import ezvcard.util.PartialDate
import org.jetbrains.anko.doAsync
import org.jetbrains.anko.uiThread
import java.io.IOException
import java.io.StringReader
import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.Future

class JournalItemActivity : BaseActivity(), Refreshable {
    private var journalEntity: JournalEntity? = null
    private lateinit var account: Account
    protected lateinit var info: CollectionInfo
    private lateinit var syncEntry: SyncEntry
    private var emailInvitationEvent: Event? = null
    private var emailInvitationEventString: String? = null

    override fun refresh() {
        val data = (applicationContext as App).data

        journalEntity = JournalEntity.fetch(data, info.getServiceEntity(data), info.uid)
        if (journalEntity == null || journalEntity!!.isDeleted) {
            finish()
            return
        }

        account = intent.extras!!.getParcelable(ViewCollectionActivity.EXTRA_ACCOUNT)!!
        info = journalEntity!!.info

        title = info.displayName

        setJournalEntryView(findViewById(R.id.journal_list_item), info, syncEntry)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.journal_item_activity)

        supportActionBar!!.setDisplayHomeAsUpEnabled(true)

        info = intent.extras!!.getSerializable(Constants.KEY_COLLECTION_INFO) as CollectionInfo
        syncEntry = intent.extras!!.getSerializable(KEY_SYNC_ENTRY) as SyncEntry

        refresh()

        val viewPager = findViewById<View>(R.id.viewpager) as ViewPager
        viewPager.adapter = TabsAdapter(supportFragmentManager, this, info, syncEntry)

        val tabLayout = findViewById<View>(R.id.tabs) as TabLayout
        tabLayout.setupWithViewPager(viewPager)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        if (emailInvitationEvent != null) {
            menuInflater.inflate(R.menu.activity_journal_item, menu)
        }
        return true
    }

    fun allowSendEmail(event: Event?, icsContent: String) {
        emailInvitationEvent = event
        emailInvitationEventString = icsContent
        invalidateOptionsMenu()
    }

    fun sendEventInvite(item: MenuItem) {
        val intent = EventEmailInvitation(this, account).createIntent(emailInvitationEvent!!, emailInvitationEventString!!)
        startActivity(intent)
    }

    private class TabsAdapter(fm: FragmentManager, private val context: Context, private val info: CollectionInfo, private val syncEntry: SyncEntry) : FragmentPagerAdapter(fm) {

        override fun getCount(): Int {
            // FIXME: Make it depend on info type (only have non-raw for known types)
            return 2
        }

        override fun getPageTitle(position: Int): CharSequence? {
            return if (position == 0) {
                context.getString(R.string.journal_item_tab_main)
            } else {
                context.getString(R.string.journal_item_tab_raw)
            }
        }

        override fun getItem(position: Int): Fragment {
            return if (position == 0) {
                PrettyFragment.newInstance(info, syncEntry)
            } else {
                TextFragment.newInstance(syncEntry)
            }
        }
    }

    class TextFragment : Fragment() {

        override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
            val v = inflater.inflate(R.layout.text_fragment, container, false)

            val tv = v.findViewById<View>(R.id.content) as TextView

            val syncEntry = arguments!!.getSerializable(KEY_SYNC_ENTRY) as SyncEntry
            tv.text = syncEntry.content

            return v
        }

        companion object {
            fun newInstance(syncEntry: SyncEntry): TextFragment {
                val frag = TextFragment()
                val args = Bundle(1)
                args.putSerializable(KEY_SYNC_ENTRY, syncEntry)
                frag.arguments = args
                return frag
            }
        }
    }

    class PrettyFragment : Fragment() {
        internal lateinit var info: CollectionInfo
        internal lateinit var syncEntry: SyncEntry
        private var asyncTask: Future<Unit>? = null

        override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
            var v: View? = null

            info = arguments!!.getSerializable(Constants.KEY_COLLECTION_INFO) as CollectionInfo
            syncEntry = arguments!!.getSerializable(KEY_SYNC_ENTRY) as SyncEntry

            when (info.type) {
                CollectionInfo.Type.ADDRESS_BOOK -> {
                    v = inflater.inflate(R.layout.contact_info, container, false)
                    asyncTask = loadContactTask(v)
                }
                CollectionInfo.Type.CALENDAR -> {
                    v = inflater.inflate(R.layout.event_info, container, false)
                    asyncTask = loadEventTask(v)
                }
                CollectionInfo.Type.TASKS -> {
                    v = inflater.inflate(R.layout.task_info, container, false)
                    asyncTask = loadTaskTask(v)
                }
            }

            return v
        }

        override fun onDestroyView() {
            super.onDestroyView()
            if (asyncTask != null)
                asyncTask!!.cancel(true)
        }

        private fun loadEventTask(view: View): Future<Unit> {
            return doAsync {
                var event: Event? = null
                val inputReader = StringReader(syncEntry.content)

                try {
                    event = Event.fromReader(inputReader, null)[0]
                } catch (e: InvalidCalendarException) {
                    e.printStackTrace()
                } catch (e: IOException) {
                    e.printStackTrace()
                }

                if (event != null) {
                    uiThread {
                        val loader = view.findViewById<View>(R.id.event_info_loading_msg)
                        loader.visibility = View.GONE
                        val contentContainer = view.findViewById<View>(R.id.event_info_scroll_view)
                        contentContainer.visibility = View.VISIBLE

                        setTextViewText(view, R.id.title, event.summary)

                        setTextViewText(view, R.id.when_datetime, getDisplayedDatetime(event.dtStart?.date?.time!!, event.dtEnd?.date!!.time, event.isAllDay(), context))

                        setTextViewText(view, R.id.where, event.location)

                        val organizer = event.organizer
                        if (organizer != null) {
                            val tv = view.findViewById<View>(R.id.organizer) as TextView
                            tv.text = organizer.calAddress.toString().replaceFirst("mailto:".toRegex(), "")
                        } else {
                            val organizerView = view.findViewById<View>(R.id.organizer_container)
                            organizerView.visibility = View.GONE
                        }

                        setTextViewText(view, R.id.description, event.description)

                        var first = true
                        var sb = StringBuilder()
                        for (attendee in event.attendees) {
                            if (first) {
                                first = false
                                sb.append(getString(R.string.journal_item_attendees)).append(": ")
                            } else {
                                sb.append(", ")
                            }
                            sb.append(attendee.calAddress.toString().replaceFirst("mailto:".toRegex(), ""))
                        }
                        setTextViewText(view, R.id.attendees, sb.toString())

                        first = true
                        sb = StringBuilder()
                        for (alarm in event.alarms) {
                            if (first) {
                                first = false
                                sb.append(getString(R.string.journal_item_reminders)).append(": ")
                            } else {
                                sb.append(", ")
                            }
                            sb.append(alarm.trigger.value)
                        }
                        setTextViewText(view, R.id.reminders, sb.toString())

                        if (event.attendees.isNotEmpty() && activity != null) {
                            (activity as JournalItemActivity).allowSendEmail(event, syncEntry.content)
                        }
                    }
                }
            }
        }

        private fun loadTaskTask(view: View): Future<Unit> {
            return doAsync {
                var task: Task? = null
                val inputReader = StringReader(syncEntry.content)

                try {
                    task = Task.fromReader(inputReader)[0]
                } catch (e: InvalidCalendarException) {
                    e.printStackTrace()
                } catch (e: IOException) {
                    e.printStackTrace()
                }

                if (task != null) {
                    uiThread {
                        val loader = view.findViewById<View>(R.id.task_info_loading_msg)
                        loader.visibility = View.GONE
                        val contentContainer = view.findViewById<View>(R.id.task_info_scroll_view)
                        contentContainer.visibility = View.VISIBLE

                        setTextViewText(view, R.id.title, task.summary)

                        setTextViewText(view, R.id.where, task.location)

                        val organizer = task.organizer
                        if (organizer != null) {
                            val tv = view.findViewById<View>(R.id.organizer) as TextView
                            tv.text = organizer.calAddress.toString().replaceFirst("mailto:".toRegex(), "")
                        } else {
                            val organizerView = view.findViewById<View>(R.id.organizer_container)
                            organizerView.visibility = View.GONE
                        }

                        setTextViewText(view, R.id.description, task.description)
                    }
                }
            }
        }

        private fun loadContactTask(view: View): Future<Unit> {
            return doAsync {
                var contact: Contact? = null
                val reader = StringReader(syncEntry.content)

                try {
                    contact = Contact.fromReader(reader, null)[0]
                } catch (e: IOException) {
                    e.printStackTrace()
                }

                if (contact != null) {
                    uiThread {
                        val loader = view.findViewById<View>(R.id.loading_msg)
                        loader.visibility = View.GONE
                        val contentContainer = view.findViewById<View>(R.id.content_container)
                        contentContainer.visibility = View.VISIBLE

                        val tv = view.findViewById<View>(R.id.display_name) as TextView
                        tv.text = contact.displayName

                        if (contact.group) {
                            showGroup(contact)
                        } else {
                            showContact(contact)
                        }
                    }
                }
            }
        }


        private fun showGroup(contact: Contact) {
            val view = this.view!!

            val mainCard = view.findViewById<View>(R.id.main_card) as ViewGroup

            addInfoItem(view.context, mainCard, getString(R.string.journal_item_member_count), null, contact.members.size.toString())

            for (member in contact.members) {
                addInfoItem(view.context, mainCard, getString(R.string.journal_item_member), null, member)
            }
        }


        private fun showContact(contact: Contact) {
            val view = this.view!!
            val mainCard = view.findViewById<View>(R.id.main_card) as ViewGroup
            val aboutCard = view.findViewById<View>(R.id.about_card) as ViewGroup
            aboutCard.findViewById<View>(R.id.title_container).visibility = View.VISIBLE

            // TEL
            for (labeledPhone in contact.phoneNumbers) {
                val types = labeledPhone.property.types
                val type = if (types.size > 0) types[0].value else null
                addInfoItem(view.context, mainCard, getString(R.string.journal_item_phone), type, labeledPhone.property.text)
            }

            // EMAIL
            for (labeledEmail in contact.emails) {
                val types = labeledEmail.property.types
                val type = if (types.size > 0) types[0].value else null
                addInfoItem(view.context, mainCard, getString(R.string.journal_item_email), type, labeledEmail.property.value)
            }

            // ORG, TITLE, ROLE
            if (contact.organization != null) {
                addInfoItem(view.context, aboutCard, getString(R.string.journal_item_organization), contact.jobTitle, contact.organization?.values!![0])
            }
            if (contact.jobDescription != null) {
                addInfoItem(view.context, aboutCard, getString(R.string.journal_item_job_description), null, contact.jobTitle)
            }

            // IMPP
            for (labeledImpp in contact.impps) {
                addInfoItem(view.context, mainCard, getString(R.string.journal_item_impp), labeledImpp.property.protocol, labeledImpp.property.handle)
            }

            // NICKNAME
            if (contact.nickName != null && !contact.nickName?.values?.isEmpty()!!) {
                addInfoItem(view.context, aboutCard, getString(R.string.journal_item_nickname), null, contact.nickName?.values!![0])
            }

            // ADR
            for (labeledAddress in contact.addresses) {
                val types = labeledAddress.property.types
                val type = if (types.size > 0) types[0].value else null
                addInfoItem(view.context, mainCard, getString(R.string.journal_item_address), type, labeledAddress.property.label)
            }

            // NOTE
            if (contact.note != null) {
                addInfoItem(view.context, aboutCard, getString(R.string.journal_item_note), null, contact.note)
            }

            // URL
            for (labeledUrl in contact.urls) {
                addInfoItem(view.context, aboutCard, getString(R.string.journal_item_website), null, labeledUrl.property.value)
            }

            // ANNIVERSARY
            if (contact.anniversary != null) {
                addInfoItem(view.context, aboutCard, getString(R.string.journal_item_anniversary), null, getDisplayedDate(contact.anniversary?.date, contact.anniversary?.partialDate))
            }
            // BDAY
            if (contact.birthDay != null) {
                addInfoItem(view.context, aboutCard, getString(R.string.journal_item_birthday), null, getDisplayedDate(contact.birthDay?.date, contact.birthDay?.partialDate))
            }

            // RELATED
            for (related in contact.relations) {
                val types = related.types
                val type = if (types.size > 0) types[0].value else null
                addInfoItem(view.context, aboutCard, getString(R.string.journal_item_relation), type, related.text)
            }

            // PHOTO
            // if (contact.photo != null)
        }

        private fun getDisplayedDate(date: Date?, partialDate: PartialDate?): String? {
            if (date != null) {
                val epochDate = date.time
                return getDisplayedDatetime(epochDate, epochDate, true, context)
            } else if (partialDate != null){
                val formatter = SimpleDateFormat("d MMMM", Locale.getDefault())
                val calendar = GregorianCalendar()
                calendar.set(Calendar.DAY_OF_MONTH, partialDate.date!!)
                calendar.set(Calendar.MONTH, partialDate.month!! - 1)
                return formatter.format(calendar.time)
            }

            return null
        }

        companion object {

            fun newInstance(info: CollectionInfo, syncEntry: SyncEntry): PrettyFragment {
                val frag = PrettyFragment()
                val args = Bundle(1)
                args.putSerializable(Constants.KEY_COLLECTION_INFO, info)
                args.putSerializable(KEY_SYNC_ENTRY, syncEntry)
                frag.arguments = args
                return frag
            }

            private fun addInfoItem(context: Context, parent: ViewGroup, type: String, label: String?, value: String?): View {
                val layout = parent.findViewById<View>(R.id.container) as ViewGroup
                val infoItem = LayoutInflater.from(context).inflate(R.layout.contact_info_item, layout, false)
                layout.addView(infoItem)
                setTextViewText(infoItem, R.id.type, type)
                setTextViewText(infoItem, R.id.title, label)
                setTextViewText(infoItem, R.id.content, value)
                parent.visibility = View.VISIBLE

                return infoItem
            }

            private fun setTextViewText(parent: View, id: Int, text: String?) {
                val tv = parent.findViewById<View>(id) as TextView
                if (text == null) {
                    tv.visibility = View.GONE
                } else {
                    tv.text = text
                }
            }

            fun getDisplayedDatetime(startMillis: Long, endMillis: Long, allDay: Boolean, context: Context?): String? {
                // Configure date/time formatting.
                val flagsDate = DateUtils.FORMAT_SHOW_DATE or DateUtils.FORMAT_SHOW_WEEKDAY
                var flagsTime = DateUtils.FORMAT_SHOW_TIME
                if (DateFormat.is24HourFormat(context)) {
                    flagsTime = flagsTime or DateUtils.FORMAT_24HOUR
                }

                val datetimeString: String
                if (allDay) {
                    // For multi-day allday events or single-day all-day events that are not
                    // today or tomorrow, use framework formatter.
                    val f = Formatter(StringBuilder(50), Locale.getDefault())
                    datetimeString = DateUtils.formatDateRange(context, f, startMillis,
                            endMillis, flagsDate).toString()
                } else {
                    // For multiday events, shorten day/month names.
                    // Example format: "Fri Apr 6, 5:00pm - Sun, Apr 8, 6:00pm"
                    val flagsDatetime = flagsDate or flagsTime or DateUtils.FORMAT_ABBREV_MONTH or
                            DateUtils.FORMAT_ABBREV_WEEKDAY
                    datetimeString = DateUtils.formatDateRange(context, startMillis, endMillis,
                            flagsDatetime)
                }
                return datetimeString
            }
        }
    }

    override fun onResume() {
        super.onResume()
        refresh()
    }

    companion object {
        private val KEY_SYNC_ENTRY = "syncEntry"

        fun newIntent(context: Context, account: Account, info: CollectionInfo, syncEntry: SyncEntry): Intent {
            val intent = Intent(context, JournalItemActivity::class.java)
            intent.putExtra(ViewCollectionActivity.EXTRA_ACCOUNT, account)
            intent.putExtra(Constants.KEY_COLLECTION_INFO, info)
            intent.putExtra(KEY_SYNC_ENTRY, syncEntry)
            return intent
        }
    }
}
